lib/net/http/header.rb

```diff
@@ -806,56 +806,107 @@ def set_form_data(params, sep = '&')
+  #
+  # - A string (ignored), an IO stream opened for reading,
+  #   and an options hash, which may contain these entries:
   #
   #   - +:filename+: The name of the file to use.
   #   - +:content_type+: The content type of the uploaded file.
   #
-  # Each item is a file field or a normal field.
-  # If +value+ is a File object or the +opt+ hash has a :filename key,
-  # the item is treated as a file field.
+  #   Example:
+  #
+  #     req.set_form([['not used', file, {filename: "other-filename.foo"}]]
+  #
+  # The various forms may be mixed:
+  #
+  #   req.set_form(['foo', %w[bar 1], ['not used', file]])
+  #
+  # <b>Argument +params+ As a Hash</b>
+  #
+  # When +params+ is a hash,
+  # each of its entries is a name/value pair that defines a field:
+  #
+  # - The name is a string.
+  # - The value may be:
+  #
+  #   - +nil+.
+  #   - Another string.
+  #   - An IO stream opened for reading.
+  #
+  # Examples:
+  #
+  #   # Nil-valued fields.
+  #   req.set_form({foo: nil, bar: nil, baz: nil})
+  #   # String to be included in request body: 'foo&bar&baz'
+  #
+  #   # String-valued fields.
+  #   req.set_form({foo: 0, bar: 1, baz: 2})
+  #   # String to be include in request body: 'foo=0&bar=1&baz=2'
   #
-  # If Transfer-Encoding is set as chunked, this sends the request using
-  # chunked encoding. Because chunked encoding is HTTP/1.1 feature,
-  # you should confirm that the server supports HTTP/1.1 before using
-  # chunked encoding.
+  #   # IO-valued field.
+  #   File.write('t.tmp', 'Ruby is cool.')
+  #   file = File.open('t.tmp')
+  #   req.set_form({not_used: file})
   #
-  # Example:
+  #   # Mixture of fields.
+  #   req.set_form({foo: nil, bar: 1, not_used: file})
   #
-  #    req.set_form([["q", "ruby"], ["lang", "en"]])
+  # Optional argument +enctype+ specifies the value to be given
+  # to field <tt>'Content-Type'</tt>, and must be one of:
   #
-  #    req.set_form({"f"=>File.open('/path/to/filename')},
-  #                 "multipart/form-data",
-  #                 charset: "UTF-8",
-  #    )
+  # - <tt>'application/x-www-form-urlencoded'</tt> (the default).
+  # - <tt>'multipart/form-data'</tt>;
+  #   see {RFC 7578}[https://www.rfc-editor.org/rfc/rfc7578].
   #
-  #    req.set_form([["f",
-  #                   File.open('/path/to/filename.bar'),
-  #                   {filename: "other-filename.foo"}
-  #                 ]],
-  #                 "multipart/form-data",
-  #    )
+  # Optional argument +formopt+ is a hash of options that may include
+  # the following entries:
   #
-  # See also RFC 2388, RFC 2616, HTML 4.01, and HTML5
+  # - +:boundary+: The value is the boundary string for the multipart message.
+  #   If not given, the boundary is a random string.
+  #   See {Boundary}[https://www.rfc-editor.org/rfc/rfc7578#section-4.1].
+  # - +:charset+: Value is the character set for the form submission.
+  #   Field names and values of non-file fields should be encoded with this charset.
```

Comment:
I think these are only applicable for `multipart/form-data`, we should note that.

Reply:
Noted.